src/DataStructures/Tensor/Expressions/Divide.hpp

```diff
@@ -0,0 +1,116 @@
+                "expression that evaluates to "
+                "a rank 0 tensor.");
+
+  using type = typename T1::type;
+  using symmetry = typename T1::symmetry;
+  using index_list = typename T1::index_list;
+  using args_list = typename T1::args_list;
+  static constexpr auto num_tensor_indices =
+      tmpl::size<typename T1::index_list>::value;
+  static constexpr auto op2_num_tensor_indices =
+      tmpl::size<typename T2::index_list>::value;
+  // the denominator has no indices or all time indices
+  static constexpr auto op2_multi_index =
+      make_array<op2_num_tensor_indices, size_t>(0);
+
+  Divide(T1 t1, T2 t2) : t1_(std::move(t1)), t2_(std::move(t2)) {}
+  ~Divide() override = default;
+
+  /// \brief Return the value of the component of the quotient tensor at a given
+  /// multi-index
+  ///
+  /// \param result_multi_index the multi-index of the component of the quotient
+  //// tensor to retrieve
+  /// \return the value of the component in the quotient tensor at
+  /// `result_multi_index`
+  SPECTRE_ALWAYS_INLINE decltype(auto) get(
+      const std::array<size_t, num_tensor_indices>& result_multi_index) const {
+    return t1_.get(result_multi_index) / t2_.get(op2_multi_index);
+  }
+
+ private:
+  T1 t1_;
+  T2 t2_;
+};
+}  // namespace TensorExpressions
+
+/// \ingroup TensorExpressionsGroup
+/// \brief Returns the tensor expression representing the quotient of one tensor
+/// expression over another tensor expression that evaluates to a rank 0 tensor
+///
+/// \details
+/// `t2` must be an expression that, when evaluated, would be a rank 0 tensor.
+/// For example, if `R` and `S` are Tensors, here is a non-exhaustive list of
+/// some of the acceptable forms that `t2` could take:
+/// - `R()`
+/// - `R(ti_A, ti_a)`
+/// - `(R(ti_A, ti_B) * S(ti_a, ti_b))`
+/// - `R(ti_t, ti_t) + 1.0`
+///
+/// \param t1 the tensor expression numerator
+/// \param t2 the rank 0 tensor expression denominator
+template <typename T1, typename T2, typename... Args2>
+SPECTRE_ALWAYS_INLINE auto operator/(
+    const TensorExpression<T1, typename T1::type, typename T1::symmetry,
+                           typename T1::index_list, typename T1::args_list>& t1,
+    const TensorExpression<T2, typename T2::type, typename T2::symmetry,
+                           typename T2::index_list, tmpl::list<Args2...>>& t2) {
+  return TensorExpressions::Divide<T1, T2, Args2...>(~t1, ~t2);
+}
+
+/// \ingroup TensorExpressionsGroup
+/// \brief Returns the tensor expression representing the quotient of a tensor
+/// expression over a `double`
+///
+/// \note The implementation instead uses the operation, `t * (1.0 / number)`
+///
+/// \param t the tensor expression operand of the quotient
+/// \param number the `double` operand of the quotient
+/// \return the tensor expression representing the quotient of a tensor
+/// expression and a `double`
+template <typename T>
+SPECTRE_ALWAYS_INLINE auto operator/(
```

Comment:
Would the inverse of this function also be useful? E.g. to write `1. / lapse()` as an expression. Related, do you intend to add `Pow` as a TE as well?

Reply:
I think `1.0 / lapse()` would be good to include in this PR. I'll push it up as a fixup.

Regarding `pow`, I've been thinking about this as well. @nilsdeppe do you agree that we should also support `pow`? I can submit this in a separate PR.

Reply:
The reason for not having `1.0 / lapse()` (so far) is to avoid people accidentally doing things like `1.0 / lapse() * shift(ti_I)`. The operation is certainly valid so if we have a lot of `1.0 /` cases we can add support. Note that currently you can do `1.0 / get(lapse)` to produce an inverse.

I haven't look at `pow` in a while. It might be supported in the same way that `square` is

Reply:
Ok, does this mean you would like me to remove the support for `1.0 / ScalarTE`?

Reply:
Wait and see what @nilsleiffischer has to say. We did explicitly not add it for the reason stated above :)

Reply:
Looking at `pow` in `src/Utilities/ConstantExpressions.hpp`([here](https://github.com/sxs-collaboration/spectre/blob/761f78545421d1dd6db157cb2e0e751b22d0c650/src/Utilities/ConstantExpressions.hpp#L159)), I do believe it should already work for positive integral powers, but I don't think `pow` will work for negative integral powers without either:
- supporting `double / ScalarTE<DataType>` generally or
- defining a new `pow` overload that returns `NumberAsExpression(double) / pow<-N>(TensorExpression)`, where `N` is a negative number

Separately, I don't think fractional powers will work until an overload of `cmath`'s `pow` is added that accepts `double` powers.

Reply:
I have no strong opinion on whether or not to allow `1.0 / lapse()`. I feel like the user will always manage to write inefficient code if they want to, and these kind of restrictions sometimes get in the way of writing reasonable code (e.g. "I just want to write 1 / lapse^7, why do I have to get() the underline DataVector for that?"). I'd merge the PR as-is, but won't object if @nilsdeppe requests we don't add the `double / ScalarTE` operation for now.

I think it's more important to solve the issue related to `square` and `pow` where the same sub-expression is evaluated multiple times, **but not in this PR**. Details below.

---

Do I understand correctly that `square(R(ti_i) * S(ti_I))` will evaluate the contraction twice? That seems quite dangerously expensive, and isn't what a user might intuitively expect (the `square` being "outside" the contraction implies it is applied after the contraction). Does the expression actually work, since it effectively reads `R(ti_i) * S(ti_I) * R(ti_i) * S(ti_I)`, which has inconsistent indices?

I don't yet understand well enough how the current implementation handles memory. Is the full TE essentially one Blaze expression, i.e. a series of arithmetic operations involving DataVectors such as `2. * x + y`, per free index? Does Blaze allocate memory for subexpressions like `2. * x` or does it compute the full expression just by using the output buffer? By extension, I'm assuming the current TE implementation allocates no intermediate memory, but computes each free index separately. E.g. `R(ti_i) * S(ti_I) * T(ti_j)` evaluates `R(ti_i) * S(ti_I)` dim-times, right? In this case the sub-expression `R(ti_i) * S(ti_I)` could be buffered in a component of the output, used to fill the other components, and then multiplied by the component of `T`. Should we make it up to the user to manage memory like that and re-use subexpressions by evaluating them explicitly, and perhaps emit a warning or even a static-assert if sub-expressions are repeated? Or can we build some level of optimization into the TEs, similar to Blaze's intra-statement optimization https://bitbucket.org/blaze-lib/blaze/wiki/Intra-Statement%20Optimization? I think for these decisions and future optimizations it's important to work with real-life use cases, so it's super useful that you're applying TEs for the CCZ4 system now. Here's an example from the XCTS system:

```cpp
/// g_ik g_jl R^ij S^kl
void fully_contract(const gsl::not_null<Scalar<DataVector>*> result,
                    const gsl::not_null<DataVector*> buffer1,
                    const gsl::not_null<DataVector*> buffer2,
                    const tnsr::II<DataVector, 3>& tensor1,
                    const tnsr::II<DataVector, 3>& tensor2,
                    const tnsr::ii<DataVector, 3>& metric) noexcept {
  get(*result) = 0.;
  for (size_t i = 0; i < 3; ++i) {
    for (size_t j = 0; j < 3; ++j) {
      *buffer1 = metric.get(i, 0) * tensor1.get(0, j);
      *buffer2 = metric.get(j, 0) * tensor2.get(i, 0);
      for (size_t k = 1; k < 3; ++k) {
        *buffer1 += metric.get(i, k) * tensor1.get(k, j);
        *buffer2 += metric.get(j, k) * tensor2.get(i, k);
      }
      get(*result) += *buffer1 * *buffer2;
    }
  }
}
```

I would very much like to write it as a TE, but will it be as efficient as this implementation? Note that this function actually constitutes a significant fraction of the runtime for a binary black-hole initial-data solve!

Reply:
okay, since the work is already done for `double / Scalar` let's leave it in. I agree that people will inevitably write inefficient code...

The optimization you're looking for in the traces is different from what Blaze does. Blaze commutes operations but doesn't create temporaries. The only way I can think of simplifying the example you gave is if TEs can allocate in the middle, which I think opens up a huge can of bugs + performance issues. You can write it as 2 exps:
```cpp
evaluate<>(scalar, S(ti_I) * R(ti_i));
evaluate<ti_j>(vector, scalar() * T(ti_j));
```
I prefer this level of control since it makes it clear where the allocations are happening. It's slightly less convenient, but I think still much better than the for loops :)

Reply:
I'm just commenting to say yes, `square(R(ti_i) * S(ti_I))` will evaluate the inner product twice and yes, `R(ti_i) * S(ti_I) * T(ti_j)` evaluates `R(ti_i) * S(ti_I)` dim-times because the expression tree has no intermediates (basically is a long blaze expression for `DataVector`) and recurses through the full tree for each unique LHS multi-index. I think I agree with both of you that avoiding this would involve the need for buffers to store intermediate values.

Reply:
And for `pow`, `cube`, and `square`, I think the repeated computation in something like `square(R(ti_I, ti_i))` could be avoided by defining a new overload for `square`, `cube`, and `pow` for TEs and define a `Pow` class (like I think Nils Fischer mentioned), and then that `Pow::get` could just use the cmath `pow`on the result of the contraction.

Let me know if this is something we're interested in or not. If so, then would we want positive, negative, and fractional powers all supported?

Reply:
Because of the implicit memory allocations I mentioned, I think this would be a bad idea to implement.

Reply:
I'm not sure I follow, so I'm going to include here a rough idea of what I was thinking:
```
#include <cmath>

template <typename T>
struct Pow {
  ...

  SPECTRE_ALWAYS_INLINE decltype(auto) get(
      const std::array<size_t, num_tensor_indices>& multi_index) const {
    return pow(t_.get(multi_index), power_);
  }

private:
  T t_;
  const double power_;
};

template <typename T, typename X, typename Symm, typename IndexList,
          typename... Args>
SPECTRE_ALWAYS_INLINE auto pow(
    const TensorExpression<T, X, Symm, IndexList, tmpl::list<Args...>>& t,
    const double power) {
  return TensorExpressions::Pow<T, Args...>(~t, power);
}
```

This doesn't fix repeated the repeated calculations over `j` in something like `R(ti_j) * pow(S(ti_I) * T(ti_i), 4.0)`, as we still compute `pow(S(ti_I) * T(ti_i), 4.0)` `j` times instead of once. However, I think this *does* fix repeated calculations in `pow(S(ti_I) * T(ti_i), 4)` that would happen if we kept things as is and used `ConstantExpressions.hpp`'s `pow` ([here](https://github.com/sxs-collaboration/spectre/blob/761f78545421d1dd6db157cb2e0e751b22d0c650/src/Utilities/ConstantExpressions.hpp#L159)), as this latter `pow` will compute the contraction `4` times and multiply them together instead of computing once and raising to the 4th power. In addition, this latter `pow` only accepts integral powers.

I don't see any implicit allocations / temporary buffers in a setup like this, but please let me know if I'm missing something!

Reply:
I think you're right that your suggestion avoids an intermediate allocations. It should call into blaze::pow (https://bitbucket.org/blaze-lib/blaze/wiki/Vector%20Operations#!pow). Then it also supports floating-point powers or indeed other vectors / ScalarTE (component-wise powers). The only issue I see is how to avoid that people use the constexpr `pow`. I suppose we could add a TE overload, but that's currently not ideal because Utilities can't depend on DataStructures. @nilsdeppe?

Reply:
Yea, but then people should just use `pow(a, b)` instead of `pow<b>(a)` if that's what they want.

Floating point powers are slower to compute than integral powers. SpEC specifically has an `ipow` for this, though this might have change recently :shrug:

I think the way to avoid recomputing things here would be to prohibit `square`, `cube`, `pow<>` with anything that isn't a `Scalar<DataVector>` TE. That is, if you try `square(S(ti_I) * T(ti_i))` you get a "this function is deleted". C++ has the `= delete` syntax that might work here (I haven't looked at it in a few years so I don't remember all the rules)

Reply:
@nilsdeppe Ok, makes sense regarding the floating points. I think the only hangup I have here is that this sounds like the same issue we discussed in this comment thread and [here](https://github.com/sxs-collaboration/spectre/pull/3549#discussion_r716913951) for scalar division, where the repeated calculation due to "bad TE writing" isn't unique to the `pow` operation (of course, while the issue exists elsewhere, it doesn't mean we can't "remove" it here). I think the use case I'm thinking about is similar to what Nils Fischer said above, if you just wanted to be able to do this, where `pow` is a new overload that doesn't do the contraction `4` times:
```
TensorExpressions::evaluate(scalar, pow<4>(S(ti_I) * T(ti_i)));
```
or this:
```
TensorExpressions::evaluate(scalar, pow<4>(S(ti_t) + T(ti_t, ti_t)));
```

so then people have the ability, but we enforce separate `evaluate` calls like [your example](https://github.com/sxs-collaboration/spectre/pull/3549#discussion_r720591179) here if someone tries to do something like this, which we don't want:
```
TensorExpressions::evaluate<ti_j>(result, R(ti_j) * pow<4>(S(ti_I) * T(ti_i)));
```

Do you think the added flexibility warrants supporting this or is it more of a cost than it's worth?

Reply:
I don't think special casing `pow<>` to behave as `pow()` is a good idea. That's sneaking things around behind the user's back. In the end all this boils down to that we have no good performance measurements so it's all just hypothesizing. It seems like my idea of disabling cases that are always suboptimal (eg `pow<N>(anything not Scalar<DataVector>)`) is not desired so then we just have to police it in code review.

Reply:
Do I understand correctly that your suggestion is this:

- Disable `pow<>(TE)`, so users don't accidentally repeat costly tensor expressions. Enable only `pow<>(TensorAsExpression<scalar stuff>)` so people _can_ write `square(lapse())` to get a TE, but _cannot_ write `square(R(ti_i) * S(ti_I))`.
- Add @macedo22's suggestion for `pow(TE, double)` above (https://github.com/sxs-collaboration/spectre/pull/3549#discussion_r721870784).

Please correct this list if I'm misrepresenting your suggestion.

Reply:
Yes, that's what I think we should do.

Reply:
Does this problem also occur with Blaze expressions?

Reply:
@nilsleiffischer @nilsdeppe Sounds good :+1: 

@wthrowe Just acknowledging that I see your question, but I don't know the answer. So, I will let one of the Nils take that one!